docs/git-guide.md

```diff
@@ -0,0 +1,1052 @@
+
+```
+$ git diff topic...master
+```
+
+## Get out of trouble
+
+Git is a powerful yet complex version control system. Even for contributors experienced at using version control, it can be confusing. The good news is that nearly all Git actions add information to the Git database, rather than removing it. As such, it's hard to make Git perform actions that you can't undo. However, git can't undo what it doesn't know about, so it's a good practice to frequently commit your changes and freqently push your commits to your remote repository.
+
+### Undo a merge commit
+
+A merge commit happens when the last commit on your current branch is not a direct ancestor of the branch you are trying to merge in. This happens quite often in a busy project like Zulip where there are many contributors because upstream/zulip will have new commits while you're working on a feature or bugfix. In order for Git to merge your changes and the changes that have occured on zulip/upstream since you first started your work, it must perform a three-way merge and create a merge commit.
+
+However, Zulip uses a forked-repo, rebase-oriented workflow without merge commits.
+
+A merge commit is usually created when you've run `git pull` or `git merge`. You'll know you're creating a merge commit if you're prompted for a commit message and the default is something like this:
+
+```
+Merge branch 'master' of https://github.com/zulip/zulip
+
+# Please enter a commit message to explain why this merge is necessary,
+# especially if it merges an updated upstream into a topic branch.
+#
+# Lines starting with '#' will be ignored, and an empty message aborts
+# the commit.
+```
+
+And the first entry for `git log` will show something like:
+
+```
+commit e5f8211a565a5a5448b93e98ed56415255546f94
+Merge: 13bea0e e0c10ed
+Author: Christie Koehler <ck@christi3k.net>
+Date:   Mon Oct 10 13:25:51 2016 -0700
+
+    Merge branch 'master' of https://github.com/zulip/zulip
+```
+
+To undo a merge commit, first run `git reflog` to identify the commit you wanto to roll back to:
+
+```
+$ git reflog
+
+e5f8211 HEAD@{0}: pull upstream master: Merge made by the 'recursive' strategy.
+13bea0e HEAD@{1}: commit: test commit for docs.
+```
+
+Reflog output will be long. The most recent git refs will be listed at the top. In the example above `e5f8211 HEAD@{0}:` is the merge commit made automatically by `git pull` and `13bea0e HEAD@{1}:` is the last commit I made before running `git pull`, the commit that I want to rollback to.
+
+Once you'd identified the ref you want to revert to, you can do so with [git reset][gitbook-reset]:
+
+```
+$ git reset --hard 13bea0e
+HEAD is now at 13bea0e test commit for docs.
+```
+
+**Important:** `git reset --hard <commit>` will discard all changes in your working directory and index since the commit you're resetting to with *<commit>*. If you need to keep any changes that are in your working directory or that you have commmited, use `git reset --merge <commit>` instead.
```

Comment:
I'd maybe add an extra word or two about how `git reset --hard` is the main way to lose work.